src/Chapter3.hs

```diff
@@ -1006,6 +1142,189 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+-- Couldn't make it simpler :(
+data KnightAction
+  = KnightAttack
+  | Drink
+  | Cast
+  deriving (Show)
+
+data MonsterAction
+  = MonsterAttack
+  | Run
+  deriving (Show)
+
+data Knight = Knight
+  { knightHealth :: Int
+  , knightAttack :: Int
+  , knightDefense :: Int
+  , knightActions :: [KnightAction]
+  , baseKnightHealth :: Int
+  , baseKnightHealthIncrease :: Int
+  , baseKnightDefense :: Int
+  , baseKnightDefenseIncrease :: Int
+  } deriving (Show)
+
+data Monster = Monster
+  { monsterHealth :: Int
+  , monsterAttack :: Int
+  , monsterActions :: [MonsterAction]
+  } deriving (Show)
+
+data ActionType
+  = GenericAttack
+  | SpecialAction
+  deriving (Show)
+
+getFirstKnightAction :: Knight -> KnightAction
+getFirstKnightAction Knight { knightActions = (x:_) } = x
+getFirstKnightAction _ = error "Forgot to 'cycle' knight actions"
+
+getFirstMonsterAction :: Monster -> MonsterAction
+getFirstMonsterAction Monster { monsterActions = (x:_) } = x
+getFirstMonsterAction _ = error "Forgot to 'cycle' monster actions"
+
+getFollowingKnightActions :: Knight -> [KnightAction]
+getFollowingKnightActions Knight { knightActions = (_:xs) } = xs
+getFollowingKnightActions _ = error "Forgot to 'cycle' knight actions"
+
+getFollowingMonsterAction :: Monster -> [MonsterAction]
+getFollowingMonsterAction Monster { monsterActions = (_:xs) } = xs
+getFollowingMonsterAction _ = error "Forgot to 'cycle' monster actions"
+
+
+class Fighter a where
+  nextAction :: a -> ActionType
+  processSpecial :: a -> a
+  processAttack :: a -> Int
+  proceed :: a -> a
+  getHealth :: a -> Int
+  takeDamage :: a -> Int -> a
+
+instance Fighter Knight where
+  nextAction :: Knight -> ActionType
+  nextAction k = case getFirstKnightAction k of
+    KnightAttack -> GenericAttack
+    _            -> SpecialAction
```

Comment:
In this case, wouldn't you just lose the information about the action?

Reply:
Oh I see, what you don't return a new creature. But in that case, what is the purpose of `General` and `Specific` action if you still would need to pattern match after?

Reply:
I don't have a good answer here. It's hard for me to make guesses on what the code's gonna look like before I've written it, and this is what I came up with.
The idea is that if it's an attack, I need to update health of another fighter, and if it's special action, I need to update the acting fighter.

Reply:
I see, neat idea of separation action by to whom that should be applied 👍🏼 
Although, for attack you still need another fighters' attack number to make the damage 🙂